File: Robot_Control_Course/Chapter11/Lesson2/Chapter11_Lesson2.java

import org.ejml.simple.SimpleMatrix;

public class JointObserver2DOF {
    private SimpleMatrix Ad;   // 4x4
    private SimpleMatrix Bd;   // 4x2
    private SimpleMatrix C;    // 2x4
    private SimpleMatrix Ld;   // 4x2
    private SimpleMatrix xHat; // 4x1

    public JointObserver2DOF() {
        Ad = SimpleMatrix.identity(4);
        Ad.set(0, 1, 0.001);
        Ad.set(2, 3, 0.001);
        // ... fill remaining entries ...

        Bd = new SimpleMatrix(4, 2);
        Bd.zero();
        Bd.set(1, 0, 0.005);
        Bd.set(3, 1, 0.005);

        C = new SimpleMatrix(2, 4);
        C.zero();
        C.set(0, 0, 1.0); // q1
        C.set(1, 2, 1.0); // q2

        Ld = new SimpleMatrix(4, 2);
        Ld.set(0, 0, 0.8);
        Ld.set(1, 0, 50.0);
        Ld.set(2, 1, 0.8);
        Ld.set(3, 1, 50.0);

        xHat = new SimpleMatrix(4, 1);
        xHat.zero();
    }

    public void reset(SimpleMatrix x0Hat) {
        xHat = x0Hat.copy();
    }

    public void update(SimpleMatrix yMeas, SimpleMatrix u) {
        // yMeas: 2x1; u: 2x1
        SimpleMatrix yHat = C.mult(xHat);
        SimpleMatrix innovation = yMeas.minus(yHat);
        xHat = Ad.mult(xHat).plus(Bd.mult(u)).plus(Ld.mult(innovation));
    }

    public SimpleMatrix getStateEstimate() {
        return xHat;
    }
}

File: Tutorials/Codes/Robot_Control_Course/Chapter11/Lesson2/Chapter11_Lesson2.cpp

#include <Eigen/Dense>

using Eigen::Matrix2d;
using Eigen::Matrix4d;
using Eigen::Vector4d;
using Eigen::Vector2d;

struct JointObserver2DOF {
  Matrix4d Ad;
  Eigen::Matrix<double,4,2> Bd;
  Eigen::Matrix<double,2,4> C;
  Eigen::Matrix<double,4,2> Ld;

  Vector4d x_hat;

  JointObserver2DOF() {
    // Example numeric values (should be computed to match your robot model)
    Ad.setIdentity();
    Ad(0,1) = 0.001;
    Ad(2,3) = 0.001;
    // ... fill remaining entries with realistic discretization ...

    Bd.setZero();
    Bd(1,0) = 0.005;
    Bd(3,1) = 0.005;

    C.setZero();
    C(0,0) = 1.0;
    C(1,2) = 1.0;

    // Observer gain Ld (example numbers):
    Ld <<
      0.8,  0.0,
      50.0, 0.0,
      0.0,  0.8,
      0.0,  50.0;

    x_hat.setZero();
  }

  void reset(const Vector4d& x0_hat) {
    x_hat = x0_hat;
  }

  void update(const Vector2d& y_meas,
              const Vector2d& u_k) {
    // Discrete-time observer update:
    // x_hat(k+1) = Ad * x_hat(k) + Bd * u(k) + Ld * (y(k) - C * x_hat(k))
    Vector2d y_hat = C * x_hat;
    x_hat = Ad * x_hat + Bd * u_k + Ld * (y_meas - y_hat);
  }

  Vector4d getStateEstimate() const {
    return x_hat;
  }
};

// In a real robot controller (e.g. in a ROS control loop),
// you would instantiate JointObserver2DOF and call update() each cycle,
// using measured joint positions as y_meas and commanded torques as u_k.
